File: utils/trade_logger.py
"""
utils/trade_logger.py
Structured CSV trade log + console/file logger setup.
"""

import csv
import logging
import os
from datetime import datetime


def setup_logging(log_file: str = "logs/trading_bot.log"):
    os.makedirs("logs", exist_ok=True)
    fmt = "%(asctime)s | %(levelname)-8s | %(name)s | %(message)s"
    logging.basicConfig(
        level=logging.INFO,
        format=fmt,
        handlers=[
            logging.FileHandler(log_file),
            logging.StreamHandler(),
        ],
    )


class TradeLogger:
    FIELDS = [
        "timestamp", "base_symbol", "etf_symbol", "action",
        "shares", "entry_price", "stop_price", "take_profit_price",
        "signal_score", "z_score", "rsi_signal", "ema_signal", "vol_conf",
        "portfolio_value", "order_id",
    ]

    def __init__(self, path: str = "logs/trades.csv"):
        os.makedirs("logs", exist_ok=True)
        self.path = path
        if not os.path.exists(path):
            with open(path, "w", newline="") as f:
                csv.DictWriter(f, fieldnames=self.FIELDS).writeheader()
        self.log = logging.getLogger("TradeLogger")

    def record(self, **kwargs):
        row = {"timestamp": datetime.utcnow().isoformat()} | kwargs
        with open(self.path, "a", newline="") as f:
            writer = csv.DictWriter(f, fieldnames=self.FIELDS, extrasaction="ignore")
            writer.writerow(row)
        self.log.info("TRADE LOGGED | %s", row)

File: app/services/trading/trade_logger.rb
# frozen_string_literal: true

require "csv"

module Trading
  class TradeLogger
    FIELDS = %w[
      timestamp base_symbol etf_symbol action
      shares entry_price stop_price take_profit_price
      signal_score z_score rsi_signal ema_signal vol_conf
      portfolio_value order_id
    ].freeze

    def initialize(path = nil)
      @path = path || Rails.root.join("log/trades.csv")
      FileUtils.mkdir_p(File.dirname(@path))
      write_header unless File.exist?(@path)
    end

    def record(**kwargs)
      row = FIELDS.index_with { nil }
      row["timestamp"] = Time.current.utc.iso8601
      row.merge!(stringify_keys(kwargs))
      CSV.open(@path, "a") do |csv|
        csv << FIELDS.map { |f| row[f] }
      end
      Rails.logger.info("TRADE LOGGED | #{row.compact}")
    end

    private

    def write_header
      CSV.open(@path, "w") { |csv| csv << FIELDS }
    end

    def stringify_keys(h)
      h.transform_keys { |k| k.to_s }
    end
  end
end
